data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/s3/grouping/S3GroupIdentifierFactory.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.s3.grouping;
+
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.plugins.sink.s3.KeyGenerator;
+
+public class S3GroupIdentifierFactory {
+
+    private final KeyGenerator keyGenerator;
+
+    public S3GroupIdentifierFactory(final KeyGenerator keyGenerator) {
+        this.keyGenerator = keyGenerator;
+     }
+
+
+    public S3GroupIdentifier getS3GroupIdentifierForEvent(final Event event) {
+
+        final String groupIdentificationHash = keyGenerator.generateKeyForEvent(event, false);
```

Comment:
I think this would have to be in `ExpressionEvaluator` since it would also need to run the expressions to create the values of the Map.

Reply:
Or are you thinking that expression evaluator would support a function to extract all of the `${}` sections out of a format expression, and then each of these individual expressions would need to be run separately to get the values? Or can we just have expression evaluator return the full Map result?

Reply:
> Or are you thinking that expression evaluator would support a function to extract all of the ${} sections out of a format expression, and then each of these individual expressions would need to be run separately to get the values?

This is what I was thinking. And they should come out in a normalized manner. That is, all have the leading `/`. This way you always get a consistent value.

> Or can we just have expression evaluator return the full Map result?

This could be an option as well. But, it seems a little more specific to this use-case. Getting the list of keys may be more extensible.

Reply:
>  This is what I was thinking. And they should come out in a normalized manner. That is, all have the leading /. This way you always get a consistent value.

So in this case, the S3 sink would have to explicitly check for keys and other expressions, and would have to check the Event itself for the values and run the expression evaluation functions individually as well